File: exercise/ts/day01/tests/food.spec.ts
import {v4 as uuidv4} from 'uuid';
import {Food} from "../src/food";

const ExpirationDate = new Date(2023, 11, 1);
const Inspector = uuidv4();

const NotFreshDate = new Date(ExpirationDate);
NotFreshDate.setDate(NotFreshDate.getDate() + 7);

const FreshDate = new Date(ExpirationDate);
FreshDate.setDate(FreshDate.getDate() - 7);

describe('Food', () => {
    test.each([
        [true, Inspector, NotFreshDate],
        [false, Inspector, FreshDate],
        [true, null, FreshDate],
        [false, null, NotFreshDate],
        [false, null, FreshDate]
    ])('is Not Edible If Not Fresh %s, %s, %s', (approvedForConsumption, inspectorId, now) => {
        const food = new Food(ExpirationDate, approvedForConsumption, inspectorId);
        expect(food.isEdible(() => now)).toBeFalsy();
    });

    test('is Edible When Fresh', () => {
        const food = new Food(ExpirationDate, true, Inspector);
        expect(food.isEdible(() => FreshDate)).toBeTruthy();
    });
});

File: solution/scala/day01/src/test/scala/FoodSpec.scala
import java.time.LocalDate
import java.time.LocalDate.of
import java.util.UUID
import java.util.UUID.randomUUID


class FoodSpec extends munit.FunSuite {

  private val expirationDate = of(2023, 12, 1)
  private val inspector = randomUUID()
  private val notFreshDate = expirationDate.plusDays(7)
  private val freshDate = expirationDate.minusDays(7)

  test("not edible food") {
    val res = notEdibleFood
      .forall {
        case (bool, maybeUuid, now) => !Food(expirationDate, bool, maybeUuid).isEdible(now)
      }
    assert(res)
  }

  test("edible food") {
    assert(Food(expirationDate, true, Some(inspector)).isEdible(freshDate))
  }

  def notEdibleFood: List[(Boolean, Option[UUID], LocalDate)] = List(
    (true, Some(inspector), notFreshDate),
    (false, Some(inspector), freshDate),
    (true, None, freshDate),
    (false, None, notFreshDate),
    (false, None, freshDate)
  )
}
